kasa/klapprotocol.py

```diff
@@ -0,0 +1,477 @@
+        # make mypy happy, this should never be reached..
+        raise SmartDeviceException("Query reached somehow to unreachable")
+
+    async def _execute_query(self, request: str, retry_count: int) -> Dict:
+        timeout = aiohttp.ClientTimeout(total=self.timeout)
+
+        async with aiohttp.ClientSession(
+            cookie_jar=self.jar, timeout=timeout
+        ) as session:
+            if not self.handshake_done or self.handshake_session_expired():
+                try:
+                    await self.perform_handshake(session)
+
+                except AuthenticationException as auex:
+                    _LOGGER.debug(
+                        "Unable to complete handshake for device %s, authentication failed",
+                        self.host,
+                    )
+                    raise auex
+
+            # Check for mypy
+            if self.encryption_session is not None:
+                payload, seq = self.encryption_session.encrypt(request.encode())
+
+            url = f"http://{self.host}/app/request"
+
+            response_status, response_data = await self.session_post(
+                session, url, params={"seq": seq}, data=payload
+            )
+
+            msg = f"at {datetime.datetime.now()}.  Host is {self.host}, Retry count is {retry_count}, Sequence is {seq}, Response status is {response_status}, Request was {request}"
+            if response_status != 200:
+                if _LOGGER.isEnabledFor(logging.ERROR):
+                    _LOGGER.error("Query failed after succesful authentication " + msg)
+                # If we failed with a security error, force a new handshake next time.
+                if response_status == 403:
+                    self.handshake_done = False
+                    self.authentication_failed = True
+                    raise AuthenticationException(
+                        f"Got a security error from {self.host} after handshake completed {self.discovery_data}"
+                    )
+                else:
+                    raise SmartDeviceException(
+                        f"Device {self.host} responded with {response_status} to request with seq {seq}"
+                    )
+            else:
+                if _LOGGER.isEnabledFor(logging.DEBUG):
+                    _LOGGER.debug("Query posted " + msg)
+
+                self.authentication_failed = False
+
+                # Check for mypy
+                if self.encryption_session is not None:
+                    decrypted_response = self.encryption_session.decrypt(response_data)
+
+                json_payload = json_loads(decrypted_response)
+
+                _LOGGER.debug(
+                    "%s << %s",
+                    self.host,
+                    _LOGGER.isEnabledFor(logging.DEBUG) and pf(json_payload),
+                )
+
+                return json_payload
+
+    async def close(self) -> None:
+        """Close the protocol.  Does nothing for this implementation."""
+        pass
```

Comment:
```suggestion
```

Reply:
Mmm, with session sharing, this is probably the place where to close the session :-)

Reply:
Is this PR working for Tapo P100 plugs? I tried discovering but it seems to not identify the klap protocol:
`kasa --target 10.42.0.130 discover --show-unsupported
Discovering devices on 10.42.0.130 for 3 seconds
Found unsupported device (tapo/unknown encryption): Unsupported device 10.42.0.130: {'result': {'device_id': '8fd258e33a4651cf8c28607690551a7f', 'owner': 
'A8086AC66FC7E2305040C2F5C302C097', 'device_type': 'SMART.TAPOPLUG', 'device_model': 'P100(UK)', 'ip': '10.42.0.130', 'mac': '40-ED-00-18-BC-04', 'is_support_iot_cloud': 
True, 'obd_src': 'tplink', 'factory_default': False, 'mgt_encrypt_schm': {'is_support_https': False, 'encrypt_type': 'KLAP', 'http_port': 80, 'lv': 2}}, 'error_code': 0}

Found 0 devices
Found 1 unsupported devices`

Reply:
No, tapo support is completely unrelated. I have created #499 which supports plugs for testing purposes, but it's undecided if that'll ever go into the library. But let's keep offtopic away from this PR , thanks.

Reply:
I've removed the `pass` but not sure what you mean about closing the session.  The ClientSession object is a context manager and the with statement lives in the query method so it should be closed.

Reply:
Ah, here it was. So basically, #537 follows the pattern where the session is re-used and not created for every query separately :-)

Reply:
Now resolved.  The AsyncClient remains connected until `protocol.close()`